packages/tokenbank_v1/contracts/modules/transfers/QuotaTransfers.sol

```diff
@@ -16,11 +16,259 @@
+    QuotaStore     public quotaStore;
+    WhitelistStore public whitelistStore;
+
+    uint public pendingExpiry;
+
+    mapping (address => mapping(bytes32 => uint)) pendingTransactions;
+
+    event PendingTxCreated (address indexed wallet, bytes32 indexed txid, uint timestamp);
+    event PendingTxExecuted(address indexed wallet, bytes32 indexed txid, uint timestamp);
+
+    constructor(
+        PriceProvider  _priceProvider,
+        SecurityStore  _securityStore,
+        QuotaStore     _quotaStore,
+        WhitelistStore _whitelistStore,
+        uint _pendingExpiry
+        )
+        public
+        SecurityModule(_securityStore)
+    {
+        priceProvider = _priceProvider;
+        quotaStore = _quotaStore;
+        whitelistStore = _whitelistStore;
+        pendingExpiry = _pendingExpiry;
+    }
+
+    function staticMethods()
+        public
+        pure
+        returns (bytes4[] memory methods)
+    {
+        methods = new bytes4[](1);
+        methods[0] = this.isPendingTxValid.selector;
+    }
+
+    function transferToken(
+        address            wallet,
+        address            token,
+        address            to,
+        uint               amount,
+        bytes     calldata data,
+        bool               enablePending
+        )
+        external
+        nonReentrant
+        onlyWhenWalletUnlocked(wallet)
+        returns (bytes32 pendingTxId)
+    {
+        bytes32 txid = keccak256(
+            abi.encodePacked(
+                "__TRANSFER__",
+                wallet,
+                token,
+                to,
+                amount,
+                keccak256(data)
+            )
+        );
+
+        bool foundPendingTx = authorizeWalletOwnerAndPendingTx(wallet, txid);
+
+        if (whitelistStore.isWhitelisted(wallet, to)) {
+            transferInternal(wallet, token, to, amount, data);
+            return bytes32(0);
+        }
+
+        uint valueInCNY = priceProvider.getValueInCNY(token, amount);
+        if (quotaStore.checkAndAddToSpent(wallet, valueInCNY)) {
+            transferInternal(wallet, token, to, amount, data);
+            return bytes32(0);
+        }
+
+        if (!foundPendingTx && enablePending) {
+            pendingTxId = txid;
+            createPendingTx(wallet, pendingTxId);
+        }
+    }
+
+    function approveToken(
+        address            wallet,
+        address            token,
+        address            to,
+        uint               amount
+        )
+        external
+        nonReentrant
+        onlyFromWalletOwner(wallet)
+        onlyWhenWalletUnlocked(wallet)
+    {
+        if (whitelistStore.isWhitelisted(wallet, to)) {
+            approveInternal(wallet, token, to, amount);
+            return;
+        }
+
+        uint allowance = ERC20(token).allowance(wallet, to);
+        if (allowance >= amount) {
+            approveInternal(wallet, token, to, amount);
+            return;
+        }
+
+        allowance -= amount;
```

Comment:
I think this needs to be `amount -= allowance;` to have the delta increased value to use in the following calculation.